### PR TITLE
Add r-base40, r-base41; Use readline8

### PR DESCRIPTION
**Pull request by babayoshihiko**
r-base40 and r-base41 are added.
Dependency updated from readline6 and readline7 to readline8.

There is a LAPACK-related error during the test process for r-base40. This seems a known bug. r-base41's corresponding test has been disabled.
https://stat.ethz.ch/pipermail/r-help/2020-May/467080.html

### CODE REVIEW

**Comment by dmacks** [comment]
Can r-base40 be patched to disable the failing test? I assume this is reg-tests-1d, and comparing tests/reg-tests-1d.R in the two versions makes it seem easy to do. Alternately, why do we even need r-base40 if you're already also supplying r-base41?

It's unusual, but possible, that both libpcre1 and libpcre2 are used. I see where libpcre2 is used, but I can't find any place that libpcre1's headers or library files are used. Could you double-check it?

**Comment by dmacks** [comment]
The Puma patch is no longer needed on either r40 or r41. I did not check back with the older r3* to see at what point it became a no-op.

**Comment by nieder** [comment]
gcc9 is currently failing on 10.15 and 11.3 (seg fault during a linking step). I have absolutely no idea on what can fix it.
But gcc11 (#846) seems to be working everywhere I've tested so far (10.14, 10.14.5, 10.15, and 11.0; 11.3 build going). Do we want to switch r-base4 to gcc11?
Also, is there a need for r-base40? Not against it, but seems superfluous if we're getting r-base41 at the same time.

**Comment by dhomeier** [comment]
Could build both on 10.14.5 with gcc9 and gcc11 alike, but r-base40-shlibs failed validation due to a wrong compatibility version:
`/sw/opt/Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libR.dylib (compatibility version 4.0.0, current version 4.0.5)`

r-base40 also had two test failures, with both compilers:

> make[3]: Entering directory '/private/var/tmp/fink.build/r-base40-4.0.5-1/R-4.0.5/tests'
running code in 'array-subset.R' ... OK
running code in 'reg-tests-1a.R' ... OK
running code in 'reg-tests-1b.R' ... OK
running code in 'reg-tests-1c.R' ... OK
running code in 'reg-tests-1d.R' ...make[3]: *** [Makefile.common:108: reg-tests-1d.Rout] Error 1
>...
> running code in 'reg-plot.R' ... OK
  comparing 'reg-plot.pdf' to './reg-plot.pdf.save' ... OK
running code in 'reg-S4-examples.R' ... OK
running code in 'reg-BLAS.R' ... OK
make[3]: Leaving directory '/private/var/tmp/fink.build/r-base40-4.0.5-1/R-4.0.5/tests'
make[2]: *** [Makefile.common:294: test-Reg] Error 2
make[2]: Leaving directory '/private/var/tmp/fink.build/r-base40-4.0.5-1/R-4.0.5/tests'
make[1]: *** [Makefile.common:173: test-all-basics] Error 1
make[1]: Target 'check' not remade because of errors.
make[1]: Leaving directory '/private/var/tmp/fink.build/r-base40-4.0.5-1/R-4.0.5/tests'
> make: *** [Makefile:306: check] Error 2
>  execution of /tmp/fink.PA1iW failed, exit code 2
>  execution of /tmp/fink.x7TKh failed, exit code 2
> phase testing: r-base40-4.0.5-1 failed

r-base41 had a single failure, but non-fatal (`make check` returned successfully):
> running code in 'reg-BLAS.R' ... OK
make[3]: Leaving directory '/private/var/tmp/fink.build/r-base41-4.1.0-1/R-4.1.0/tests'
make[3]: Entering directory '/private/var/tmp/fink.build/r-base41-4.1.0-1/R-4.1.0/tests'
During startup - Warning messages:
1: Setting LC_CTYPE failed, using "C" 
2: Setting LC_COLLATE failed, using "C" 
3: Setting LC_TIME failed, using "C" 
4: Setting LC_MESSAGES failed, using "C" 
5: Setting LC_MONETARY failed, using "C" 
running code in 'reg-tests-3.R' ... OK
  comparing 'reg-tests-3.Rout' to './reg-tests-3.Rout.save' ... OK
During startup - Warning messages:
1: Setting LC_CTYPE failed, using "C" 
2: Setting LC_COLLATE failed, using "C" 
3: Setting LC_TIME failed, using "C" 
4: Setting LC_MESSAGES failed, using "C" 
5: Setting LC_MONETARY failed, using "C" 
running code in 'reg-examples3.R' ... OK
  comparing 'reg-examples3.Rout' to './reg-examples3.Rout.save' ... OK
running tests of plotting Latin-1
  expect failure or some differences if not in a Latin-1 or UTF-8 locale
running code in 'reg-plot-latin1.R' ... FAILED

In addition there were lots of the LC warnings in both packages, which could be avoided by running `LC_ALL make ...` in the Compile and TestScripts.

In summary, looks good and +1 for switching this to gcc11, if that builds better on the other distributions.

**Comment by dmacks** [comment]
r-base40's failure in reg-tests-1d.R is the LAPACK-related one that can be ignored.

**Comment by babayoshihiko** [comment]
I am working with gcc11 and also other issues (libpcre1, reg-tests-1d.R bug, Puma patch, 4.1.0 compile option).

There is a big change between 4.0 and 4.1. This change may affect cran- packages significantly. In this sese, 4.1. is more like 5.0 (or 4.0 is more like 3.7) to my personally. As users often use R in several proje4ct teams and may want to use a specific version for such projects, I want to keep 4.0.

**Comment by nieder** [comment]
@babayoshihiko thanks for the 4.0/4.1 explanation. No problem with keeping 4.0 if it's useful.

File: 10.9-libcxx/stable/main/finkinfo/sci/r-base40.info [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+  <<
+  Files: <<
+  Library/Frameworks/R.framework/Versions/%type_raw[rversion]/Resources/etc/Renviron
+  Library/Frameworks/R.framework/Versions/%type_raw[rversion]/Resources/etc/ldpaths
+  Library/Frameworks/R.framework/Versions/%type_raw[rversion]/Resources/lib
+  Library/Frameworks/R.framework/Versions/%type_raw[rversion]/Resources/library
+  Library/Frameworks/R.framework/Versions/%type_raw[rversion]/Resources/modules
+  <<
+  Shlibs: <<
+  %p/Library/Frameworks/R.framework/Versions/%type_raw[rversion]/Resources/lib/libR.dylib 4.0.5 %n (>= 4.0.5-1)
```
Comment (dhomeier) [inline 10.9-libcxx/stable/main/finkinfo/sci/r-base40.info:217]:
```suggestion
  %p/Library/Frameworks/R.framework/Versions/%type_raw[rversion]/Resources/lib/libR.dylib 4.0.0 %n (>= 4.0.5-1)
```
see comment on the validation error. The actual starting version for that would probably be `(>= 4.0.0-1)`, but since there is no packaged version for that,  it should be correct the way above.

File: 10.9-libcxx/stable/main/finkinfo/sci/r-base40.info [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+  SED=/usr/bin/sed AWK=/usr/bin/awk GREP=/usr/bin/grep
+<<
+CompileScript: <<
+  #!/bin/sh -efv
+ 
+  . %p/sbin/fink-buildenv-helper.sh
+
+  export F77=%p/bin/gfortran-fsf-9
+  export FC=%p/bin/gfortran-fsf-9
+
```
Comment (dhomeier) [inline 10.9-libcxx/stable/main/finkinfo/sci/r-base40.info:85]:
```suggestion
    export LC_ALL=C
 
```
to get rid of the LANG-warnings

File: 10.9-libcxx/stable/main/finkinfo/sci/r-base40.info [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+InfoTest: <<
+  TestScript: <<
+  #!/bin/sh -efv
+    # don't leak from already-installed package
+    unset R_HOME
+    # set TZ to avoid https://bugs.r-project.org/bugzilla3/show_bug.cgi?id=17307
+    export TZ=`date +%%Z`
+    
+    # reg-test-1d.R checks error messages, but the correct answer is English only
+    export LANGUAGE=en
```
Comment (dhomeier) [inline 10.9-libcxx/stable/main/finkinfo/sci/r-base40.info:103]:
```suggestion
    export LANGUAGE=en LC_ALL=C
```

File: 10.9-libcxx/stable/main/finkinfo/sci/r-base40.info [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+    make -j1 -k check || exit 2
+  <<
+<<
+InstallScript: <<
+  #!/bin/sh -ex
+  # prefix=%i/Library/Frameworks
+  make install DESTDIR=%d
```
Comment (dhomeier) [inline 10.9-libcxx/stable/main/finkinfo/sci/r-base40.info:110]:
```suggestion
  LC_ALL=C make install DESTDIR=%d
```

File: 10.9-libcxx/stable/main/finkinfo/sci/r-base41.info [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+  SED=/usr/bin/sed AWK=/usr/bin/awk GREP=/usr/bin/grep
+<<
+CompileScript: <<
+  #!/bin/sh -efv
+ 
+  . %p/sbin/fink-buildenv-helper.sh
+
+  export F77=%p/bin/gfortran-fsf-9
+  export FC=%p/bin/gfortran-fsf-9
+
```
Comment (dhomeier) [inline 10.9-libcxx/stable/main/finkinfo/sci/r-base41.info:84]:
```suggestion
    export LC_ALL=C
 
```

File: 10.9-libcxx/stable/main/finkinfo/sci/r-base41.info [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+  popd
+  fink-package-precedence --depfile-ext='\.d' --prohibit-bdep r-base,%n,%n-dev .
+<<
+InfoTest: <<
+  TestScript: <<
+  #!/bin/sh -efv
+    # don't leak from already-installed package
+    unset R_HOME
+    # set TZ to avoid https://bugs.r-project.org/bugzilla3/show_bug.cgi?id=17307
+    export TZ=`date +%%Z`
```
Comment (dhomeier) [inline 10.9-libcxx/stable/main/finkinfo/sci/r-base41.info:99]:
```suggestion
    export TZ=`date +%%Z`
    export LC_ALL=C
```

File: 10.9-libcxx/stable/main/finkinfo/sci/r-base41.info [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+    make -j1 -k check || exit 2
+  <<
+<<
+InstallScript: <<
+  #!/bin/sh -ex
+  # prefix=%i/Library/Frameworks
+  make install DESTDIR=%d
```
Comment (dhomeier) [inline 10.9-libcxx/stable/main/finkinfo/sci/r-base41.info:106]:
```suggestion
  LC_ALL=C make install DESTDIR=%d
```

**Review by dhomeier — CHANGES_REQUESTED** [review]
Might be my non-en.US user setup, but as mentioned I got these
```
During startup - Warning messages:
1: Setting LC_CTYPE failed, using "C" 
2: Setting LC_COLLATE failed, using "C" 
3: Setting LC_TIME failed, using "C" 
4: Setting LC_MESSAGES failed, using "C" 
5: Setting LC_MONETARY failed, using "C" 
  base
```
on every compilation in some `make` stages.

**Comment by dmacks** [comment]
The gccNN used to build r-baseXX propagates into many rmod packages. Essentially whatever compiler version is used when r-base itself is first packaged is the one we're stuck with for that rXX version. Definitely good to use the latest one at the time to avoid needing to retain as much support for old compilers far into the future.

File: 10.9-libcxx/stable/main/finkinfo/sci/r-base41.info (hunk 5 above)
Comment (babayoshihiko) [reply 10.9-libcxx/stable/main/finkinfo/sci/r-base41.info:84]: I suppose gfortran shoudl also be 11.

**Comment by dmacks** [comment]
All my concerns are resolved with r40 and r41 (both built cleanly on 10.13). The readline upgrades look good by eye. Awaiting @nieder's #846 to be merged.

**Comment by dmacks** [comment]
gcc11 has landed. The only detail I see in the r-base* world is that it would be cleaner to have a ```Type:gcc``` variable to centralize the version. That would reduce the diff among r-base*.info, and more importantly be where to place a very clear note that this should not be changed due to propagation into the mods.

**Comment by nieder** [comment]
I've tried and tried to get gcc9 to build on macOS12, but have not succeeded and I doubt I will (supposedly possible but I don't have the skills). So any r-base (and its rmods) currently using gcc9 will end up needing to be dist restricted.
I know gcc9 currently works on 10.14.5 and earlier and gcc11 works on 12.1, but don't know which gcc releases work on 10.15-11.6.

**Comment by dmacks** [comment]
@dhomeier's concerns/requests seem to have been handled by df7a15d

**Comment by nieder** [comment]
gcc9/10 work through 10.14.5, maybe 10.15. Most likely they will fail on macOS 11.
I'm OK with pushing.

**Comment by dhomeier** [comment]
Thanks, both build with the updated gcc11 on 10.13.6 and 10.14.6, i.e. gcc11 so far works on High Sierra as well!
Passing almost all tests; this one also seems to behave as it should:
```
running tests of plotting Latin-1
  expect failure or some differences if not in a Latin-1 or UTF-8 locale
running code in 'reg-plot-latin1.R' ... FAILED
```
The r-base40 tests on 10.13 without a TeX installation produced a single failure:

**Review by dhomeier — APPROVED** [review]

**Comment by dmacks** [comment]
@dhomeier just to clarify, those two test failures are diagnostic messages but non-fatal to the overall package building? Or do we need to disable some tests (or have a TestDepends or set a language env-var)?
